Add crypto module

vladchicos opened the pull request:
Add hkdf_extract
Add hkdf_extract unittest

vladchicos committed:
Add crypto module

Add hkdf_extract
Add hkdf_extract unittest

Geod24 commented:
Why the switch to Nix ? The action is doing a good job on its own.

Geod24 commented on dub.sdl:
```diff
@@ -5,3 +5,5 @@ copyright "Copyright © 2022, Vlăduț Chicoș"
 
 targetPath "build"
+
+dependency "openssl" version="~>3.2.2"
```
```suggestion
dependency "openssl" version="~>3.2"
```

vladchicos committed:
Add HKDF-Expand-Label
Add HKDF-Expand-Label unittest
Add packet encryption

Geod24 commented on src/quic/crypto.d:
```diff
@@ -0,0 +1,68 @@
+module quic.crypto;
```
Every module should come with documentation

Geod24 commented on src/quic/crypto.d:
```diff
@@ -0,0 +1,159 @@
+    kctx = EVP_KDF_CTX_new(kdf);
+    EVP_KDF_free(kdf);
+
+    if (EVP_KDF_CTX_set_params(kctx, p) < 1)
+        handleErrors;
+    if (EVP_KDF_derive(kctx, initial_secret.ptr, 32, p) < 1)
+        handleErrors;
```
Why is every error fatal ?

vladchicos replied:
Another approach that I think of would be to print the error queue and then return an error value. Is this a better idea?

vladchicos commented:
I used two different approaches for hkdf-expand and hkdf-extract. That is because in the meantime I figured out how to use EVP_PKEY_CTX related methods(compatible with openssl 1.1.1). Should I keep the one using EVP_KDF functions(added in openssl 3.0) or the other one? For the one with EVP_KDF I had to make(and fill with data) my own ossl_param structs due to the fact that deimos openssl does not bind yet to all the required methods(see this example for reference : https://wiki.openssl.org/index.php/EVP_Key_Derivation). Should I make a PR into deimos with them? Any kind of suggestion/solution is welcome :) . Thanks!

vladchicos committed:
Add a better way of signaling errors
Add a decryption function
hkdf_extract uses EVP_PKEY_CTX api

vladchicos committed:
Add encrypt/decrypt_packet unittest

vladchicos committed:
Selective imports

vladchicos committed:
Add AES-ECB encryption and decryption

vladchicos committed:
Add public/private key generator
Add generateSharedKey
Add digest
More suggestive name

Fix function naming in the unittest

Fix error handling

vladchicos committed:
Add signature validation

vladchicos committed:
Add unittests
generateSharedKey with templates
Initialize pctx for shared key derivation

vladchicos committed:
Add a test
